### DISCUSSION TITLE
Modules support — am I doing this right?

### DISCUSSION THREAD

**Discussion by intelfx** [arighi/virtme-ng#203] [Q&A] [answered]
(Someone pointed me at virtme-ng just today, so I'm not entirely familiar with how it works, apologies.)

Let's say that instead of using virtme-ng to build a fresh new minimal kernel, I have an already configured and built kernel tree in `$PWD` with a full distro config, modules and everything. I want to run that kernel with full support for in-tree and out-of-tree modules (dkms, ...) alike.

I didn't see an option to specifically point `vng` at a built kernel tree. According to the help, it's either `vng -b` to build in $PWD, or `vng -r /path/to/bzImage/or/dir` to run a standalone binary. So I just ran `vng -r .`, and it seemed to autodiscover the modules and do the right thing, even with a `/lib/modules/$(uname -r)/build` symlink back to the source tree.

Is it actually supported to run vng like that (i.e. point `--run` at a built kernel tree and expect that it will grab not just the kernel binary but also the modules and headers) or is it all just a happy coincidence?

**Comment by arighi** [comment, marked as answer]
vng is automatically using the modules from the build dir, basically it installs all the modules in your target kernel build dir ($PWD in this case) in `.virtme_mods/` and it will bind mount this dir in `/lib/modules/$kernel_version`, so inside the vng session you'll be able to load external modules normally.

Keep in mind that this will replace the original `/lib/modules/$kernel_version` on the host, so if you have out-of-tree dkms modules you won't be able to see them with your recompiled kernel. But you can see them if you run the host's kernel (using `vng -r`).

Also note that `vng -r .` is equivalent to simply running `vng` (default is to run the kernel from the current working dir).

**Reply by intelfx** [reply]
So if the argument to `-r` (or the implied `-r .`) points to a build tree, vng _is_ supposed to not just discover the kernel binary (as described in the help), but also discover the built modules and do the right thing™?

**Reply by arighi** [reply]
Correct, it's supposed to be smart enough to detect both the kernel binary, the corresponding modules and do the right thing(tm) - most of the times. :)
